make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+        """
+        Builds the PRG in prg_as_list. The PRG of a MultiIntervalNode is a concatenation of the PRG of its children
+        """
+        for child in self.children:
+            child.preorder_traversal_to_build_prg(prg_as_list, delim_char)
+
+
+class MultiClusterNode(RecursiveTreeNode):
+    """
+    Represent nodes that are clustered horizontally, i.e. in sequence clusters
+    """
+    def __init__(self, nesting_level: int, alignment: MSA, parent: Optional["RecursiveTreeNode"],
+                 prg_builder: "PrgBuilder", cluster_subalignments: SubMSAs):
+        super().__init__(nesting_level, alignment, parent, prg_builder, cluster_subalignments)
+        assert not self.is_leaf(), "MultiClusterNodes should never be leaves"
+
+    def preorder_traversal_to_build_prg(self, prg_as_list: List[str], delim_char: str = " "):
+        """
+        Builds the PRG in prg_as_list. The PRG of a MultiClusterNode consists of opening a site,
+        putting each child as an allele and closing the site
+        """
+        site_num = self.prg_builder.get_next_site_num()
+        prg_as_list.extend(f"{delim_char}{site_num}{delim_char}")
+
+        for child_index, child in enumerate(self.children):
+            site_num_to_separate_alleles = (
+                (site_num + 1) if (child_index < len(self.children) - 1) else site_num
+            )
+            child.preorder_traversal_to_build_prg(prg_as_list, delim_char)
+            prg_as_list.extend(
+                f"{delim_char}{site_num_to_separate_alleles}{delim_char}"
+            )
+
+
+class UpdateError(Exception):
+    pass
+
+
+class LeafNode(RecursiveTreeNode):
+    """
+    Represent nodes that are never clustered in either direction.
+    These nodes are the only ones that can get updated and indexed.
+    """
+    def __init__(self, nesting_level: int, alignment: MSA, parent: Optional["RecursiveTreeNode"],
+                 prg_builder: "PrgBuilder"):
+        super().__init__(nesting_level, alignment, parent, prg_builder, [])
+        assert self.is_leaf(), f"Leaf node ({self}) is not a leaf"
+
+        # update-related attributes
+        self.new_sequences: Set[str] = set()
+        self.indexed_PRG_intervals: Set[Tuple[int, int]] = set()
+
+    def preorder_traversal_to_build_prg(self, prg_as_list: List[str], delim_char: str = " ", do_indexing=True):
+        """
+        Builds the PRG in prg_as_list. The PRG of a leaf node is the sequences themselves it represents
+        """
+        expanded_sequences = SequenceExpander.get_expanded_sequences_from_MSA(self.alignment)
+
+        single_seq = len(expanded_sequences) == 1
+        if single_seq:
+            start_index = len(prg_as_list)
+            prg_as_list.extend(expanded_sequences[0])
+            end_index = len(prg_as_list)
+            if do_indexing:
+                self.prg_builder.update_PRG_index(start_index, end_index, node=self)
```

Comment:
this is repeated with lines 176-177 below, could write a single dedented copy of it at the end

Reply:
The one in lines 176-177 is inside a loop, is done several times, not only once

Reply:
yep sorry!